### PR TITLE
feat(build-cli): New command check:prApproval

### PR DESCRIPTION
**Pull request by tylerbutler**
Adds a new flub command, check:prApproval, which is used to check if a PR in the repo is approved by a GitHub team or list of users.

This command is then used in a new workflow that runs on release branch PRs to check that a member of the FluidFramework-ReleaseApprovers team has approved the PR. We'll then make that a required check for release branch PRs.

Finally, the CODEOWNERS file is updated to require approval from FluidFramework-ReleaseApprovers for changes to the workflow itself.

The command can be run locally with a GitHub PAT, which is how I tested. I also tested against a test branch on my fork. However, none of these tests is an exact analog to the real Actions environment.

Because of how Actions work, this workflow must be merged to main before it can be further tested using branches.

[AB#8814](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/8814)

### CODE REVIEW

File: .github/workflows/required-reviewers-release.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+          which flub
+          flub --help
+          flub commands
+
+      - name: Check PR approval
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          # This command will fail with an error if the PR is not approved, which
+          # will cause
```
Comment (Josmithr) [inline .github/workflows/required-reviewers-release.yml:64]: Seems like the note is incomplete

File: build-tools/packages/build-cli/src/commands/check/prApproval.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+			description: "The PR number to check.",
+			required: true,
+			min: 1,
+		}),
+		team: Flags.string({
+			description:
+				"The team whose membership should be checked. The team must be in the same GitHub organization as the repo. Only the team name should be provided - the org is inferred from the repo details.",
+			exactlyOne: ["team", "approvers"],
+		}),
+		approvers: Flags.string({
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/commands/check/prApproval.ts:57]: Must the entire list of "approvers" here approve the PR for the check to succeed? Or just one of them?

File: build-tools/packages/build-cli/src/library/githubRest.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	 */
+	owner: string;
+
+	/**
+	 * The name of the repo.
+	 */
+	repo: string;
+
+	/**
+	 * A GitHub access token.
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/library/githubRest.ts:23]: Nit: might be worth adding a `@remarks` with best practices for token use, just to be extra security cautious.

File: .github/workflows/required-reviewers-release.yml [hunk 4]
```diff
@@ -0,0 +1,68 @@
+name: Required release approval
+on:
+  pull_request_target:
+    types:
+      - opened
+
+      # This is triggered when the base branch changes; handles the case where you open a PR against one branch
+      # then change the base branch to a release branch.
+      - edited
```
Comment (Josmithr) [inline .github/workflows/required-reviewers-release.yml:9]: Is there a way to trigger this when the set of approvers has changed? Seems like we would want to re-run validation after a user approves, without needing to push changes to the branch. Or am I missing something?

File: build-tools/packages/build-cli/src/commands/check/prApproval.ts [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+	// eslint-disable-next-line import/no-internal-modules
+} from "../../library/githubRest.js";
+import { BaseCommand } from "../../library/index.js";
+
+/**
+ * This command class is used to merge two branches based on the batch size provided.
+ * It looks for the last common commit between two branches and computes the remaining commits to be merged.
+ * Later, it creates a pull request based on the batch size passed.
+ */
+export default class CheckPrApprovalCommand extends BaseCommand<
```
Comment (Josmithr) [inline build-tools/packages/build-cli/src/commands/check/prApproval.ts:24]: If we were to want to enforce multiple review requirements for the same conditions, what would the flow look like for using this (e.g., if multiple teams were required)? Looks like currently you would have to call this separately for each reviewer group? Wondering if this is a case we should consider here.

**Review by Josmithr — COMMENTED** [review]
I left some questions and suggestions, but all-in-all seems reasonable to me. As currently configured, this would immediately begin to block changes on release branches, right? If so, are we ready for that? Seems like someone on the release team should probably approve.

File: .github/workflows/required-reviewers-release.yml (hunk 4 above)
Comment (tylerbutler) [reply .github/workflows/required-reviewers-release.yml:9]: Great catch. We need to trigger this on pull_request_review as well.

File: build-tools/packages/build-cli/src/commands/check/prApproval.ts (hunk 5 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/check/prApproval.ts:24]: Yes, it only checks one group so you'd have to call it multiple times. Seems easy enough to update the command to take multiple teams, or multiple groups, or whatever, should requirements
arise. That said, doing them separately can make for a better report UX for the CI run.

File: build-tools/packages/build-cli/src/commands/check/prApproval.ts (hunk 2 above)
Comment (tylerbutler) [reply build-tools/packages/build-cli/src/commands/check/prApproval.ts:57]: Clarified. Only one is needed.

**Comment by tylerbutler** [comment]
> As currently configured, this would immediately begin to block changes on release branches, right?

No - the workflow has to be ported to a release branch to take effect. So the earliest it would "automatically" start would be after next release.

File: .github/workflows/required-reviewers-release.yml [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        with:
+          persist-credentials: false
+          submodules: false
+
+      # install and configure node, pnpm and the changeset tools
+      # release notes: https://github.com/pnpm/action-setup/releases/tag/v4.0.0
+      - uses: pnpm/action-setup@fe02b34f77f8bc703788d5817da081398fad5dd2 # ratchet:pnpm/action-setup@v4
+
+      # release notes: https://github.com/actions/setup-node/releases/tag/v4.0.3
+      - uses: actions/setup-node@1e60f620b9541d16bece96c5465dc8ee9832be0b # ratchet:actions/setup-node@v3
```
Comment (tylerbutler) [inline .github/workflows/required-reviewers-release.yml:42]:
```suggestion
      - uses: actions/setup-node@1e60f620b9541d16bece96c5465dc8ee9832be0b # ratchet:actions/setup-node@v4
```

**Review by Josmithr — APPROVED** [review]

**Review by tylerbutler — COMMENTED** [review]
test review

**Review by tylerbutler — COMMENTED** [review]
test trigger

**Review by tylerbutler — COMMENTED** [review]
test review

**Pull request merged**